### PR TITLE
Fix connection/memory leak

### PR DESCRIPTION
**Pull request by feuerbach**
Hi,

First of all, thanks for the library — we at @nstack use it with great success.

Unfortunately, we discovered that as it is right now, the library does not close the connections and leaks memory allocated to the connection thread.

There are two causes of this:

1. In order for the finalizer to run, it has to be attached to a TVar with a TVar-specific function.
2. The connection thread itself retains the connection object, and so it is never GC'd and the finalizer never runs.

You can reproduce the leak by observing the memory consumption of this simple program depending on the value of the argument:

``` haskell
import System.Environment
import qualified DBus as DBus
import Control.Monad
import System.Mem (performGC)

callDBusCommand :: IO ()
callDBusCommand = do
  _ <- DBus.connectClient DBus.System
  performGC

main = do
  n <- read . head <$> getArgs
  replicateM_ n callDBusCommand
```

This PR fixes the memory leak. It also removes the unnnecessary weak pointer on the alive ref and the unnecessary exception handler in the connection thread.

### CODE REVIEW

**Comment by Philonous** [comment]
I'm glad this code is useful for someone (I'm afraid I'll have to actually write tests now :> ) 

And thanks for the great PR!

I have a few concerns though. 

It seems to me that removing the exception handler from the worker thread means we won't notice when the thread dies (say because of a parse failure), which would result in the connection being stuck and threads waiting for method responses dangling. 

Simlarly, active methods calls (that is, those that have not yet returned or been garbage collected) should arguably keep the connection from being harvested as well, that's why https://github.com/nstack/d-bus/blob/master/src/DBus/Message.hs#L310-L324 returns an action that touches the connectionAliveRef, which would probably have to be updated to touch the gcRef instead.

**Comment by RayRacine** [comment]
Too funny just dropped regarding the same issue that I just today. 

 Is it the design intent that an explicit DBusConnection closeConnection is not required in the API? 

In my use-case using the DBus Bluez  API, I connect and establish a  number of DBus signal / property change handlers for a slew of Bluetooth BLE devices.   These handlers will fire responding to incoming property changes, device discovery, etc. and typically the handler is performing side effecting ops, say updating a device IOT database, or responding to some remote Bluetooth thermometer's temperature reading changing or whatever.  

In lieu, of attempting to maintain all the bookkeeping of a bunch of match patterns and signal/property handler registrations in some cases it is just easier to close the old connection and open a fresh one that will utilize a different set of match patterns and handlers.   I'd like to be able to explicitly close the DBusConnection, waiting for a full graceful close if necessary, so I know that no signal/property handlers are still firing doing all sorts of side effects in background until eventually GC'd.

Right now with this connection leak all the stale connection's handlers keep firing away.  When the leak is fixed however handlers will still fire until an eventual GC cleans up the connection.  But eventual could be  ... well who knows how long.

Currently the API does have the "waitFor" API, but that is for a different purpose (detection of failure for keep alive use case).  But it does allow to me wait for the eventual connection to get GC scarfed, but AFAIK there is no guarantee when, if ever, that GC occurs.

In summary, does the API not need an explicit, gracefully executing closeConnection function?  This would allow for explicit deterministic closing, as opposed to non-deterministic eventual closing.

**Comment by feuerbach** [comment]
@Philonous: thanks for the quick feedback!

> It seems to me that completely removing the exception handler from the worker thread means we won't notice when the thread dies (say because of a parse failure), which would result in the connection being stuck and threads waiting for method responses dangling.

Good catch — I didn't realize `kill` wasn't called anywhere else. I updated the PR with a fix.

It would be nice to enable warnings (`-Wall`) and fix all the existing ones, that'd catch my mistake.

I will now look into your second remark.

**Comment by Philonous** [comment]
@RayRacine I seem to remeber I had a function for this, but I can't find it. Could you please open another issue?

**Comment by feuerbach** [comment]
@Philonous I've updated the PR to incorporate your second comment.

We should also decide and clarify what the semantics of waitFor should be regarding the garbage collection.
We have two choices: waitFor can keep the connection alive or not.

I've updated `waitFor` to keep the connection alive (which is consistent with the intention of the previous version), but it means that the connection will always be alive unless an error occurs in the connection thread, or the user manually writes False to the connectionAliveRef. What do you think is the more useful one?

Whatever semantics we choose, we should clearly document it.

**Comment by Philonous** [comment]
The reason I included `waitFor` was for the case where the application exported an interface to be consumed by others. It is indeed supposed to keep the connection alive and block indefinitely.
